File: ktkernel/subsystems/pmm/pmm.cpp
#include "subsystems/pmm/pmm.h"
#include "limine/requests.h"

namespace kt_kernel
{
    constexpr auto min_block_size = 1ull << min_order;

    inline size_t get_order(size_t size)
    {
        if (size <= min_block_size)
            return 0;

        return bit_length(size - 1) - bit_length(min_block_size - 1);
    }

    inline size_t bit_length(size_t size)
    {
        if (size == 0) [[unlikely]]
            return 0;

        return (sizeof(size) * 8) - __builtin_clzll(size);
    }

    uint64_t buddy_allocator::get_buddy(uint64_t address, size_t order)
    {
        return address ^ (1ull << (order + min_order));
    }

    void buddy_allocator::add_region(uint64_t base, size_t size)
    {
        base = (base + min_block_size - 1) & ~(min_block_size - 1);
        size &= ~(min_block_size - 1);

        while (size >= min_block_size) {
            size_t order = orders - 1;
            while (order > 0 && (1ull << (order + min_order)) > size)
                order--;
            while (order > 0 && (base & ((1ull << (order + min_order)) - 1)) != 0)
                order--;

            size_t block_size = 1ull << (order + min_order);
            auto* blk = reinterpret_cast<block*>(base + hhdm_offset);
            blk->next = free_lists[order];
            free_lists[order] = blk;

            base += block_size;
            size -= block_size;
        }
    }

    void buddy_allocator::split(size_t order)
    {
        block* blk = free_lists[order];
        free_lists[order] = blk->next;

        size_t half_size = 1ull << ((order - 1) + min_order);
        auto* second = reinterpret_cast<block*>(reinterpret_cast<uint64_t>(blk) + half_size);
        second->next = free_lists[order - 1];
        blk->next = second;
        free_lists[order - 1] = blk;
    }

    uint64_t buddy_allocator::allocate_pages(size_t order)
    {
        size_t current = order;
        while (current < orders && free_lists[current] == nullptr)
            current++;

        if (current >= orders)
            return 0;

        while (current > order) {
            split(current);
            current--;
        }

        block* blk = free_lists[order];
        free_lists[order] = blk->next;

        return reinterpret_cast<uint64_t>(blk) - hhdm_offset;
    }

    void buddy_allocator::coalesce(uint64_t addr, size_t order)
    {
        while (order < orders - 1) {
            uint64_t buddy_physical = get_buddy(addr, order);
            auto* buddy_virtual = reinterpret_cast<block*>(buddy_physical + hhdm_offset);

            block** prev = &free_lists[order];
            block* curr = free_lists[order];

            bool found = false;

            while (curr) {
                if (curr == buddy_virtual) {
                    *prev = curr->next;
                    found = true;
                    break;
                }

                prev = &curr->next;
                curr = curr->next;
            }

            if (!found)
                break;

            addr = (addr < buddy_physical) ? addr : buddy_physical;
            order++;
        }

        auto* blk = reinterpret_cast<block*>(addr + hhdm_offset);
        blk->next = free_lists[order];
        free_lists[order] = blk;
    }

    void buddy_allocator::free_pages(uint64_t addr, size_t order)
    {
        coalesce(addr, order);
    }

    bool pmm::initialize()
    {
        this->hhdm_offset = limine::hhdm_request.response->offset;
        engine.hhdm_offset = this->hhdm_offset;

        for (uint64_t i = 0; i < limine::memmap_request.response->entry_count; i++) {
            auto* entry = limine::memmap_request.response->entries[i];
            if (entry->type == LIMINE_MEMMAP_USABLE) {
                engine.add_region(entry->base, entry->length);
                total_usable_pages += (entry->length / 4096);
            }
        }

        return true;
    }

    uint64_t pmm::allocate_frame()
    {
        uint64_t frame = engine.allocate_pages(0);
        if (frame)
            used_pages++;

        return frame;
    }

    uint64_t pmm::allocate_contiguous(size_t n)
    {
        if (n == 0)
            return 0;

        size_t order = get_order(n * 4096);
        uint64_t frame = engine.allocate_pages(order);

        if (frame)
            used_pages += (1ull << order);

        return frame;
    }

    void pmm::free_frame(uint64_t frame)
    {
        engine.free_pages(frame, 0);
        if (used_pages > 0)
            used_pages--;
    }

    void pmm::free_contiguous(uint64_t frame, size_t n)
    {
        size_t order = get_order(n * min_block_size);
        engine.free_pages(frame, order);
        if (used_pages >= (1ull << order))
            used_pages -= (1ull << order);
    }

    size_t pmm::get_total_memory()
    {
        return total_usable_pages * 4096;
    }

    size_t pmm::get_used_memory()
    {
        return used_pages * 4096;
    }
} // namespace kt_kernel

File: ktkernel/src/memory/pmm.rs
use spin::Mutex;

use crate::limine::{HHDM, MEMMAP};
use limine::memmap::MEMMAP_USABLE;

const MAX_ORDER: usize = 25;
const MIN_ORDER: usize = 12;
const ORDERS: usize = MAX_ORDER - MIN_ORDER + 1;
const MIN_BLOCK_SIZE: u64 = 1 << MIN_ORDER;
const PAGE_SIZE: u64 = 4096;

#[repr(C)]
struct FreeBlock {
    next: *mut FreeBlock,
}

struct BuddyAllocator {
    free_lists: [*mut FreeBlock; ORDERS],
    hhdm_offset: u64,
}

unsafe impl Send for BuddyAllocator {}

impl BuddyAllocator {
    const fn new() -> Self {
        Self {
            free_lists: [core::ptr::null_mut(); ORDERS],
            hhdm_offset: 0,
        }
    }

    fn get_buddy(&self, address: u64, order: usize) -> u64 {
        address ^ (1u64 << (order + MIN_ORDER))
    }

    fn add_region(&mut self, mut base: u64, mut size: u64) {
        base = (base + MIN_BLOCK_SIZE - 1) & !(MIN_BLOCK_SIZE - 1);
        size &= !(MIN_BLOCK_SIZE - 1);

        while size >= MIN_BLOCK_SIZE {
            let mut order = ORDERS - 1;
            while order > 0 && (1u64 << (order + MIN_ORDER)) > size {
                order -= 1;
            }
            while order > 0 && (base & ((1u64 << (order + MIN_ORDER)) - 1)) != 0 {
                order -= 1;
            }

            let block_size = 1u64 << (order + MIN_ORDER);
            let block = (base + self.hhdm_offset) as *mut FreeBlock;
            unsafe {
                (*block).next = self.free_lists[order];
            }
            self.free_lists[order] = block;

            base += block_size;
            size -= block_size;
        }
    }

    fn split(&mut self, order: usize) {
        let block = self.free_lists[order];
        unsafe {
            self.free_lists[order] = (*block).next;

            let half_size = 1u64 << ((order - 1) + MIN_ORDER);
            let second = (block as u64 + half_size) as *mut FreeBlock;
            (*second).next = self.free_lists[order - 1];
            (*block).next = second;
            self.free_lists[order - 1] = block;
        }
    }

    fn allocate_pages(&mut self, order: usize) -> u64 {
        let mut current = order;
        while current < ORDERS && self.free_lists[current].is_null() {
            current += 1;
        }

        if current >= ORDERS {
            return 0;
        }

        while current > order {
            self.split(current);
            current -= 1;
        }

        let block = self.free_lists[order];
        unsafe {
            self.free_lists[order] = (*block).next;
        }

        block as u64 - self.hhdm_offset
    }

    fn coalesce(&mut self, mut addr: u64, mut order: usize) {
        while order < ORDERS - 1 {
            let buddy_physical = self.get_buddy(addr, order);
            let buddy_virtual = (buddy_physical + self.hhdm_offset) as *mut FreeBlock;

            let mut prev = &mut self.free_lists[order] as *mut *mut FreeBlock;
            let mut curr = self.free_lists[order];
            let mut found = false;

            while !curr.is_null() {
                if curr == buddy_virtual {
                    unsafe { *prev = (*curr).next };
                    found = true;
                    break;
                }
                unsafe {
                    prev = &mut (*curr).next;
                    curr = (*curr).next;
                }
            }

            if !found {
                break;
            }

            addr = addr.min(buddy_physical);
            order += 1;
        }

        let block = (addr + self.hhdm_offset) as *mut FreeBlock;
        unsafe {
            (*block).next = self.free_lists[order];
        }
        self.free_lists[order] = block;
    }

    fn free_pages(&mut self, addr: u64, order: usize) {
        self.coalesce(addr, order);
    }
}

pub struct Pmm {
    buddy: BuddyAllocator,
    total_usable_pages: usize,
    used_pages: usize,
}

impl Pmm {
    const fn new() -> Self {
        Self {
            buddy: BuddyAllocator::new(),
            total_usable_pages: 0,
            used_pages: 0,
        }
    }

    pub fn initialize(&mut self) {
        let hhdm = HHDM.response().expect("HHDM response missing");
        let memmap = MEMMAP.response().expect("memmap response missing");

        self.buddy.hhdm_offset = hhdm.offset;

        for entry in memmap.entries() {
            if entry.type_ == MEMMAP_USABLE {
                self.buddy.add_region(entry.base, entry.length);
                self.total_usable_pages += (entry.length / PAGE_SIZE) as usize;
            }
        }
    }

    pub fn allocate_frame(&mut self) -> Option<u64> {
        let frame = self.buddy.allocate_pages(0);
        if frame != 0 {
            self.used_pages += 1;
            Some(frame)
        } else {
            None
        }
    }

    pub fn allocate_contiguous(&mut self, n: usize) -> Option<u64> {
        if n == 0 {
            return None;
        }

        let order = get_order(n as u64 * PAGE_SIZE);
        let frame = self.buddy.allocate_pages(order);

        if frame != 0 {
            self.used_pages += 1 << order;
            Some(frame)
        } else {
            None
        }
    }

    pub fn free_frame(&mut self, frame: u64) {
        self.buddy.free_pages(frame, 0);
        self.used_pages = self.used_pages.saturating_sub(1);
    }

    pub fn free_contiguous(&mut self, frame: u64, n: usize) {
        let order = get_order(n as u64 * MIN_BLOCK_SIZE);
        self.buddy.free_pages(frame, order);
        self.used_pages = self.used_pages.saturating_sub(1 << order);
    }

    pub fn total_memory(&self) -> usize {
        self.total_usable_pages * PAGE_SIZE as usize
    }

    pub fn used_memory(&self) -> usize {
        self.used_pages * PAGE_SIZE as usize
    }

    pub fn hhdm_offset(&self) -> u64 {
        self.buddy.hhdm_offset
    }
}

fn get_order(size: u64) -> usize {
    if size <= MIN_BLOCK_SIZE {
        return 0;
    }
    bit_length(size - 1) - bit_length(MIN_BLOCK_SIZE - 1)
}

fn bit_length(size: u64) -> usize {
    if size == 0 {
        return 0;
    }
    64 - size.leading_zeros() as usize
}

pub static PMM: Mutex<Pmm> = Mutex::new(Pmm::new());
